iscservice/demo/libdemo.c

```diff
@@ -0,0 +1,352 @@
+	int               n;
+	uint32_t          shift;
+	double            val;
+	FILE             *f;
+	char             *p;
+	struct mm_result  res = {};
+	struct m0_buf     buf = M0_BUF_INIT0;
+	struct m0_stob_io *stio = (struct m0_stob_io *)pdata->icp_data;
+
+	shift = m0_stob_block_shift(stio->si_obj);
+	bufvec_open(&stio->si_user, shift);
+	p = stio->si_user.ov_buf[0];
+
+	f = fmemopen(p, stio->si_count << shift, "r");
+	if (f == NULL) {
+		*rc = M0_ERR(-errno);
+		return M0_FSO_AGAIN;
+	}
+
+	/*
+	 * 1st value should go to the lbuf always, because it can be
+	 * the right cut of the last value from the previous unit.
+	 *
+	 * We count it as the 1st element in the buffer.
+	 */
+	if (fscanf(f, "%lf %n", &val, &n) < 1) {
+		fclose(f);
+		*rc = M0_ERR(-EINVAL);
+		return M0_FSO_AGAIN;
+	}
+	res.mr_lbuf.b_addr = p;
+	res.mr_lbuf.b_nob = n;
+	res.mr_nr = 1;
+
+	/* Read 1st element to compare with (2nd element in the buffer). */
+	if (fscanf(f, "%lf ", &val) < 1)
+		goto out;
+	res.mr_val = val;
+	res.mr_idx = 1;
+
+	for (i = 2; fscanf(f, "%lf %n", &val, &n) > 0 && !feof(f); i++) {
+		if (op == MIN ? val < res.mr_val :
+		                val > res.mr_val) {
+			res.mr_idx = i;
+			res.mr_val = val;
+		}
+	}
+	res.mr_nr = i;
+	/*
+	 * Last value should always go to the rbuf, because it can be
+	 * the left cut of the first value in the next unit.
+	 *
+	 * rbuf element is counted as lbuf in the next unit. It may
+	 * happen that there is no cut of element on the border. I.e.
+	 * the last element is fully present at the end of the unit and
+	 * the 1st element is fully present in the beginning of the next
+	 * unit. This case must be checked by the client code (it should
+	 * assume that there might be two elements present in the result
+	 * of gluing rbuf and lbuf of who units). And if so, the index
+	 * and the number of elements should be incremented by client.
+	 */
+	res.mr_rbuf.b_addr = p + ftell(f) - n;
+	res.mr_rbuf.b_nob = n;
+ out:
+	*rc = m0_xcode_obj_enc_to_buf(&M0_XCODE_OBJ(mm_result_xc, &res),
+				      &buf.b_addr, &buf.b_nob) ?:
+	      m0_buf_copy_aligned(out, &buf, M0_0VEC_SHIFT);
+
+	m0_buf_free(&buf);
+	fclose(f);
+
+	return M0_FSO_AGAIN;
+}
+
+int do_minmax(enum op op, struct m0_buf *in, struct m0_buf *out,
+	      struct m0_isc_comp_private *data, int *rc)
+{
+	int                res;
+	struct m0_stob_io *stio = (struct m0_stob_io *)data->icp_data;
+
+	if (stio == NULL) {
+		M0_ALLOC_PTR(stio);
+		if (stio == NULL) {
+			*rc = -ENOMEM;
+			return M0_FSO_AGAIN;
+		}
+		data->icp_data = stio;
+		res = launch_stob_io(data, in, rc);
+		if (*rc != -EAGAIN)
+			m0_free(stio);
+	} else {
+		res = compute_minmax(op, data, out, rc);
+		stio_fini(stio, stio->si_obj);
+		m0_free(stio);
+	}
+
+	return res;
+}
+
+int comp_min(struct m0_buf *in, struct m0_buf *out,
+	     struct m0_isc_comp_private *comp_data, int *rc)
+{
+	return do_minmax(MIN, in, out, comp_data, rc);
+}
+
+int comp_max(struct m0_buf *in, struct m0_buf *out,
```

Comment:
Please add detailed explanation of the function prototype.
The parameters type.
The expected and valid result value. (this is very important).

Reply:
It's described here - https://github.com/Seagate/cortx-motr/blob/4f852dfa0fd2000d7f210665f5176894b39885f8/iscservice/isc.h#L75-L85 and here - https://github.com/Seagate/cortx-motr/blob/main/doc/ISC_Service_User_Guide.rst

Maybe we should add some comments here also, indeed. Will do.

Reply:
Done.